public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts

```diff
@@ -511,3 +511,346 @@ describe('getAfterSelectorCompletions', () => {
     expect(labelFiltersInSuggestions.length).toBe(0);
   });
 });
+
+describe('IN_LOGFMT completions', () => {
+  let datasource: LokiDatasource;
+  let languageProvider: LokiLanguageProvider;
+  let completionProvider: CompletionDataProvider;
+
+  beforeEach(() => {
+    datasource = createLokiDatasource();
+    languageProvider = new LokiLanguageProvider(datasource);
+    completionProvider = new CompletionDataProvider(languageProvider, {
+      current: history,
+    });
+
+    jest.spyOn(completionProvider, 'getParserAndLabelKeys').mockResolvedValue({
+      extractedLabelKeys: ['label1', 'label2'],
+      unwrapLabelKeys: [],
+      hasJSON: true,
+      hasLogfmt: false,
+      hasPack: false,
+    });
+  });
+  it('autocompleting logfmt should return flags, parsers, pipe operations, and labels', async () => {
+    const situation: Situation = {
+      type: 'IN_LOGFMT',
+      logQuery: `{job="grafana"} | logfmt`,
+      flags: false,
+      otherLabels: [],
+    };
+    
+    expect(await getCompletions(situation, completionProvider)).toMatchInlineSnapshot(`
+      [
```

Comment:
> Here it seems like we are mixing mock completion (with mock docs) and real completion (with real docs). Why is that?

Because the mocks come from:

```
jest.mock('../../../querybuilder/operations', () => ({
  explainOperator: () => 'Operator docs',
}));
```

The reason is so tests doesn't break because we change the documentation. The usual "tests should be resilient to change".

> Also these snapshot tests are quite long (100s of lines of code). What if we just checked that it has correct labels, rather than checking everything, as that's what we care about.

The whole way we test the completion output here is very cumbersome, to say the least. After having worked multiple times on this, I'm now more inclined to use inline snapshots, as they give you a readable output of the completion in a "similar" way of using the editor and seeing what it offers. I find that with snapshots it's easier to see that the output is good, unlike https://github.com/grafana/grafana/blob/e16617dcb75fa5e54f615133b730d687885ff285/public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts#L361-L362 where the expectations are invisible.
 
We were also doing "snapshots" in the form of hardcoded json objects in places like this: https://github.com/grafana/grafana/blob/e16617dcb75fa5e54f615133b730d687885ff285/public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts#L231-L255

If I were to test this again, at the moment I would change everything to snapshots or json objects, but I'm 100% open to suggestions. Right now I can't see a way of testing completions without having to use very large strings or objects.

Reply:
I would think here to check only important things. For example documentation is not that important, but `label` and `type`.